src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -0,0 +1,978 @@
+/*
+ * This file is part of AtomVM.
+ *
+ * Copyright 2023 Winford (Uncle Grumpy) <winford@object.stream>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+ */
+
+#if !(defined(AVM_DISABLE_GPIO_PORT_DRIVER) && defined(AVM_DISABLE_GPIO_NIFS))
+
+#include <stdbool.h>
+#include <string.h>
+
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/exti.h>
+#include <libopencm3/stm32/gpio.h>
+
+#include <atom.h>
+#include <bif.h>
+#include <context.h>
+#include <defaultatoms.h>
+#include <globalcontext.h>
+#include <interop.h>
+#include <mailbox.h>
+#include <nifs.h>
+#include <scheduler.h>
+#include <sys.h>
+#include <term.h>
+// #define ENABLE_TRACE
+#include <trace.h>
+
+#include "avm_log.h"
+#include "gpio_driver.h"
+#include "stm_sys.h"
+
+#define TAG "gpio_driver"
+
+#define GPIOInvalidBank 0x0000U
+#define GPIO_MODE_OUTPUT_OD 0x4
+#define GPIO_INVALID_MODE 0xE
+#define INVALID_GPIO_OSPEED 0xE
+
+#define HIGH_ATOM globalcontext_make_atom(ctx->global, ATOM_STR("\x4", "high"))
+#define LOW_ATOM globalcontext_make_atom(ctx->global, ATOM_STR("\x3", "low"))
```

Comment:
As we discussed, this is okay on stm32 because 
1) globalcontext_make_atom is idepmpotent and will return the atom if it is in the table in O(c) time
2) stm32 does not support SMP, so no lock contenttion